cpp/src/link_prediction/similarity_impl.cuh

```diff
@@ -162,5 +163,302 @@ rmm::device_uvector<weight_t> similarity(
+        graph_view.local_vertex_partition_view());
+      auto num_invalid_vertices =
+        thrust::count_if(handle.get_thrust_policy(),
+                         source_vertices->begin(),
+                         source_vertices->end(),
+                         [vertex_partition] __device__(auto val) {
+                           return !(vertex_partition.is_valid_vertex(val) &&
+                                    vertex_partition.in_local_vertex_partition_range_nocheck(val));
+                         });
+
+      if constexpr (multi_gpu) {
+        num_invalid_vertices = cugraph::host_scalar_allreduce(
+          handle.get_comms(), num_invalid_vertices, raft::comms::op_t::SUM, handle.get_stream());
+      }
+
+      CUGRAPH_EXPECTS(num_invalid_vertices == 0,
+                      "Invalid input arguments: there are invalid input vertices.");
+    }
+
+    if (edge_weight_view) {
+      auto num_negative_edge_weights =
+        count_if_e(handle,
+                   graph_view,
+                   edge_src_dummy_property_t{}.view(),
+                   edge_dst_dummy_property_t{}.view(),
+                   *edge_weight_view,
+                   [] __device__(vertex_t, vertex_t, auto, auto, weight_t w) { return w < 0.0; });
+
+      if constexpr (multi_gpu) {
+        num_negative_edge_weights = cugraph::host_scalar_allreduce(handle.get_comms(),
+                                                                   num_negative_edge_weights,
+                                                                   raft::comms::op_t::SUM,
+                                                                   handle.get_stream());
+      }
+
+      CUGRAPH_EXPECTS(
+        num_negative_edge_weights == 0,
+        "Invalid input argument: input edge weights should have non-negative values.");
+    }
+  }
+
+  rmm::device_uvector<vertex_t> sources(0, handle.get_stream());
+
+  if (source_vertices) {
+    sources.resize(source_vertices->size(), handle.get_stream());
+    thrust::copy(handle.get_thrust_policy(),
+                 source_vertices->begin(),
+                 source_vertices->end(),
+                 sources.begin());
+  } else {
+    sources.resize(graph_view.local_vertex_partition_range_size(), handle.get_stream());
+    thrust::sequence(handle.get_thrust_policy(),
+                     sources.begin(),
+                     sources.end(),
+                     graph_view.local_vertex_partition_range_first());
+  }
+
+  if (topk) {
+    //  We can reduce memory footprint by doing work in batches and
+    //  computing/updating topk with each batch
+    rmm::device_uvector<vertex_t> top_v1(0, handle.get_stream());
+    rmm::device_uvector<vertex_t> top_v2(0, handle.get_stream());
+    rmm::device_uvector<weight_t> top_score(0, handle.get_stream());
+
+    top_v1.reserve(*topk, handle.get_stream());
+    top_v2.reserve(*topk, handle.get_stream());
+    top_score.reserve(*topk, handle.get_stream());
```

Comment:
We can defer defining/allocating these variables till right before the beginning of the while loop to minimize the variable scopes.